File: web/src/schemas/simulation.ts
import { z } from "zod";

export const NodeStatusSchema = z.object({
  name: z.string(),
  identity_hash: z.string().default(""),
  role: z.string().default("endpoint"),
  status: z.string().default("offline"),
  lat: z.number().nullable().optional(),
  lon: z.number().nullable().optional(),
  path_count: z.number().default(0),
  announce_count: z.number().default(0),
  link_count: z.number().default(0),
  uptime: z.number().default(0),
});

export const LinkStatusSchema = z.object({
  source: z.string(),
  target: z.string(),
  medium: z.string(),
});

export const SimulationStatusSchema = z.object({
  running: z.boolean().default(false),
  scenario_name: z.string().nullable().optional(),
  node_count: z.number().default(0),
  nodes: z.array(NodeStatusSchema).default([]),
  links: z.array(LinkStatusSchema).default([]),
});

export type NodeStatus = z.infer<typeof NodeStatusSchema>;
export type LinkStatus = z.infer<typeof LinkStatusSchema>;
export type SimulationStatus = z.infer<typeof SimulationStatusSchema>;

File: rnetsim/api/models/simulation.py
"""Pydantic models for simulation state and status responses."""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel


class NodeStatus(BaseModel):
    """Status of a single simulation node."""

    name: str
    identity_hash: str = ""
    role: str = "endpoint"
    status: str = "offline"  # healthy, degraded, offline, sleeping
    lat: Optional[float] = None
    lon: Optional[float] = None
    path_count: int = 0
    announce_count: int = 0
    link_count: int = 0
    uptime: float = 0


class LinkStatus(BaseModel):
    """Status of a link between two nodes."""

    source: str
    target: str
    medium: str


class SimulationStatus(BaseModel):
    """Aggregated status of the running simulation."""

    running: bool = False
    scenario_name: Optional[str] = None
    node_count: int = 0
    nodes: list[NodeStatus] = []
    links: list[LinkStatus] = []
